hazelcast/test/src/sql_test.cpp

```diff
@@ -0,0 +1,491 @@
+    }
+    static void TearDownTestSuite()
+    {
+        member_.reset();
+        member2_.reset();
+        server_factory_.reset();
+    }
+
+    portable_pojo_key key(int64_t i) {
+        return {i};
+    }
+
+    portable_pojo value(int64_t i) {
+        return portable_pojo{i};
+    }
+
+private:
+    static std::unique_ptr<HazelcastServerFactory> server_factory_;
+    static std::unique_ptr<HazelcastServer> member_;
+    static std::unique_ptr<HazelcastServer> member2_;
+};
+
+std::unique_ptr<HazelcastServerFactory> SqlTest::server_factory_{};
+std::unique_ptr<HazelcastServer> SqlTest::member_{};
+std::unique_ptr<HazelcastServer> SqlTest::member2_{};
+
+TEST_F(SqlTest, simple)
+{
+    sql::sql_statement statement(client, R"sql(
+        SELECT * FROM (VALUES ('foo', 'bar'), (NULL, 'hello')) AS X(col1, col2)
+    )sql");
+
+    sql::sql_service service = client.get_sql();
+    auto result = service.execute(statement).get();
+
+    ASSERT_TRUE(result.is_row_set());
+    EXPECT_EQ(-1, result.update_count());
+    ASSERT_TRUE(result.row_metadata());
+    ASSERT_EQ(2, result.row_metadata()->columns().size());
+    auto& column0 = result.row_metadata()->columns()[0];
+    EXPECT_EQ("col1", column0.name());
+    EXPECT_EQ(hazelcast::client::sql::sql_column_type::varchar, column0.type());
+    EXPECT_TRUE(column0.nullable());
+    auto& column1 = result.row_metadata()->columns()[1];
+    EXPECT_EQ("col2", column1.name());
+    EXPECT_EQ(hazelcast::client::sql::sql_column_type::varchar, column1.type());
+    EXPECT_FALSE(column1.nullable());
+
+    auto page_it = result.page_iterator();
+    auto const& page = *page_it;
+    ASSERT_TRUE(page.has_value());
+    auto& rows = page->rows();
+    EXPECT_EQ(2, rows.size());
+    EXPECT_EQ("foo", rows[0].get_object<std::string>(0).value());
+    EXPECT_EQ("bar", rows[0].get_object<std::string>(1).value());
+    EXPECT_FALSE(rows[1].get_object<std::string>(0).has_value());
+    EXPECT_EQ("hello", rows[1].get_object<std::string>(1).value());
+}
+
+TEST_F(SqlTest, statement_with_params)
+{
+    sql::sql_service service = client.get_sql();
+    auto result = service
+                    .execute("SELECT CAST(? AS VARCHAR), CAST(? AS VARCHAR)",
+                             123456,
+                             -42.73)
+                    .get();
+
+    ASSERT_TRUE(result.is_row_set());
+    EXPECT_EQ(-1, result.update_count());
+
+    auto page = *result.page_iterator();
+    ASSERT_TRUE(page);
+
+    auto& rows = page->rows();
+    EXPECT_EQ(1, rows.size());
+    EXPECT_EQ("123456", rows[0].get_object<std::string>(0).value());
+    EXPECT_EQ("-42.73", rows[0].get_object<std::string>(1).value());
+}
+
+TEST_F(SqlTest, exception)
+{
+    sql::sql_service service = client.get_sql();
+    EXPECT_THROW(service.execute("FOO BAR", 42).get(),
+                 hazelcast::client::sql::hazelcast_sql_exception);
+}
+
+// ported from Java SqlBasicClientTest.testSelect
+TEST_F(SqlTest, select)
```

Comment:
Is this complete? Where is the query part?

Reply:
check it now.